File: znai-reactjs/src/doc-elements/api/ApiParameters.js
/*
 * Copyright 2020 znai maintainers
 * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from "react";
import ApiParameter from "./ApiParameter";

import "./ApiParameters.css";

export default function ApiParameters({
                                        parameters,
                                        title,
                                        references,
                                        nestedLevel,
                                        small,
                                        noWrap,
                                        wide,
                                        collapsible,
                                        collapsed,
                                        parentWidth = 0,
                                        next,
                                        prev,
                                        elementsLibrary
                                      }) {
  const [userDrivenCollapsed, setUserDrivenCollapsed] = useState(collapsed);

  const isExpanded = !nestedLevel && parameters.length === 1 && parameters[0].children;

  const renderedParameters = userDrivenCollapsed ? null :
    (parameters || []).map(p => <ApiParameter key={p.name}
                                              anchorId={p.anchorId}
                                              name={p.name}
                                              type={p.type}
                                              isExpanded={isExpanded}
                                              children={p.children}
                                              description={p.description}
                                              nestedLevel={nestedLevel}
                                              references={references}
                                              elementsLibrary={elementsLibrary} />);

  const isNested = nestedLevel > 0;
  const className = "znai-api-parameters" +
    (isNested ? " nested" : "") +
    (small ? " small" : "") +
    (noWrap ? " no-wrap" : "");

  const style = { marginLeft: -parentWidth };

  const rendered = (
    <div className={className} style={style}>
      <ApiParametersTitle title={title} nestedLevel={nestedLevel} collapsible={collapsible} collapsed={userDrivenCollapsed} collapseToggle={collapseToggle}/>
      {renderedParameters}
    </div>
  );

  if (!isNested) {
    const containerClass = "znai-api-parameters-wrapper " +
      (wide ? "wide" : "content-block") +
      (next?.type === "ApiParameters" ? " no-bottom-margin" : "") +
      (prev?.type === "ApiParameters" ? " no-top-margin" : "")

    return (
      <div className={containerClass}>
        {rendered}
      </div>
    )
  }

  return rendered

  function collapseToggle() {
    setUserDrivenCollapsed(prev => !prev)
  }
}

function ApiParametersTitle({ title, nestedLevel, collapsible, collapsed, collapseToggle}) {
  if (!title || nestedLevel > 0) {
    return null;
  }

  return (
    <div className="znai-api-parameters-title-cell">
      {collapsible && (<div className="znai-api-parameters-collapse-toggle"
                            onClick={collapseToggle}>{collapsed ? "+" : "-"}</div>)}

      <div className="znai-api-parameters-title">
        {title}
      </div>
    </div>
  );
}

File: znai-core/src/main/java/org/testingisdocumenting/znai/extensions/api/ApiParameters.java
/*
 * Copyright 2020 znai maintainers
 * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.testingisdocumenting.znai.extensions.api;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.toList;

public class ApiParameters {
    private final ApiParameter root;
    private String example;

    public ApiParameters(String anchorPrefix) {
        root = new ApiParameter(ApiParametersAnchors.sanitizeAnchorId(anchorPrefix),
                "root", new ApiLinkedText("root"), Collections.emptyList(), "");
    }

    public ApiParameter addInFront(String name, ApiLinkedText type, List<Map<String, Object>> description, String textForSearch) {
        return root.addInFront(name, type, description, textForSearch);
    }

    public ApiParameter add(String name, ApiLinkedText type, List<Map<String, Object>> description, String textForSearch) {
        return root.add(name, type, description, textForSearch);
    }

    public void add(ApiParameter parameter) {
        root.add(parameter);
    }

    public ApiParameter find(String name) {
        return root.find(name);
    }

    public String getAnchorId() {
        return root.getAnchorId();
    }

    public String getExample() {
        return example;
    }

    public void setExample(String example) {
        this.example = example;
    }

    public ApiParameters withoutTopLevel() {
        ApiParameters result = new ApiParameters(root.getAnchorId());
        result.example = example;
        if (root.getChildren().isEmpty()) {
            return result;
        }

        ApiParameter firstChild = root.getChildren().get(0);
        firstChild.getChildren().forEach(result::add);

        return result;
    }

    public ApiParameter getRoot() {
        return root;
    }

    public boolean isEmpty() {
        return root.getChildren().isEmpty();
    }

    public Map<String, Object> toMap() {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("parameters", root.getChildren().stream().map(ApiParameter::toMap).collect(toList()));
        result.put("example", example);

        return result;
    }

    public List<String> collectAllAnchors() {
        return root.getChildren().stream()
                .flatMap(child -> child.collectAllAnchors().stream())
                .collect(toList());
    }

    public String combinedTextForSearch() {
        return root.getChildren().stream()
                .map(ApiParameter::combinedTextForSearch)
                .collect(Collectors.joining(" "));
    }
}
